### ISSUE TITLE
gzip falure on checkpoint

### ISSUE THREAD

**Issue by MrPrezident** [dmtcp/dmtcp#370]
The environment that I'm in uses a wrapper script for a lot of the common unix utilities such as gzip. i.e. $PATH points to a version of gzip that is really a shell script wrapper that ultimately resolves to the real platform dependent gzip executable. The problem is that the wrapper script relies on some environment variables, but when gzip is called from dmtcp, all of the environment variables are gone. i.e. when I try to create a checkpoint, dmtcp calls gzip, which then fails due to undefined environment variables and the checkpoint hangs. It's as if gzip was invoked like this: env -i /path/to/gzip

Why do the env vars get wiped during the gzip call, and is there anything way to fix this?

**Comment by jiajuncao** [comment]
Could you try `--no-gzip` option of dmtcp_launch as a solution? This gets rid of the call to gzip.

**Comment by MrPrezident** [comment]
The App that I use generates checkpoint files that are generally multiple GB in size, so I don't think --no-gzip would be a good idea.

**Comment by jiajuncao** [comment]
What's the error message?

**Comment by rohgarg** [comment]
@MrPrezident: I haven't looked at our code that invokes gzip very closely, but it's likely that there's some piece there that's wiping out the environment variables. I'll try to look at it later today and get back to you.

**Comment by MrPrezident** [comment]
@jiajuncao Here is the error message:

```
/p/flow/iwutilities/bin/gzip: line 55: -MFile::Basename: command not found
/p/flow/iwutilities/bin/gzip: line 67: /.linux50/gzip: No such file or directory
/p/flow/iwutilities/bin/gzip: line 67: exec: /.linux50/gzip: cannot execute: No such file or directory
```
 
When I get these errors, the checkpoint file ends up being of size zero. These are the exact same errors that I get when I do "env -i /p/flow/iwutilities/bin/gzip". Also, I tried creating a fake gzip script that wrote out all of the env vars into a file when executed. Here is the result:

```
PWD=/nfs/project/work/dmtcp-dmtcp-30d2101
SHLVL=1
_=/bin/env
```

@rohgarg Thanks!

I noticed that there is a dmtcp doc about environment variables in the git repo: dmtcp/doc/environment-vars.txt

> After initialization, environment variables are maintained solely in
> user space, and are ignored by the kernel.

Does this account for the problem?

**Comment by rohgarg** [comment]
No, that's talking about how the environment variables are preserved as part of the checkpoint image. This provides a target application a consistent view of its environment across checkpoint and restart. If you read the same paragraph further it explains why presenting a consistent view is important.

>For example, a shell program may cache its own copy of the environment
variables prior to checkpoint.  It would be bad if environment variables
asynchronously appeared and disappeared as part of checkpoint/restart.

The issue you are seeing relates to how, at checkpoint time, DMTCP searches for a gzip binary in the PATH, and forks and execs gzip to compress the checkpoint data. The gzip-related code is in `src/ckptserializer.cpp`, if you are interested. It's possible that either the search for gzip isn't working correctly or the call to fork-and-exec gzip strips off some environment variables.

**Comment by MrPrezident** [comment]
I figured out the problem. Line 365 in src/ckptserializer.cpp should be changed from this:

`    _real_execve(extcomp_args[0], extcomp_args, NULL);`

to this:

`    _real_execv(extcomp_args[0], extcomp_args);`

execve with NULL passed in clears the env vars. execv (without the 'e') leaves the env vars alone. I made the change and tested it and it works.

**Comment by rohgarg** [comment]
Ah! I think this is the right place, but if I'm not wrong, we end up inheriting the `LD_PRELOAD` variable with your proposed fix. We don't want gzip to run under DMTCP.

**Comment by MrPrezident** [comment]
You could also do this (or some variant):
`_real_execve(extcomp_args[0], extcomp_args, environ);`

**Comment by rohgarg** [comment]
I guess, we just need to remove the `LD_PRELOAD` env. var. from the environment before the exec call, similar to what's done in `src/dmtcp_nocheckpoint.c`.

**Comment by MrPrezident** [comment]
Thanks @rohgarg! Do you have an ETA on when you think you would be able to update that?

**Comment by rohgarg** [comment]
Sorry, I just started looking at the code, and the fix is as simple as what you had suggested earlier. :-)

Earlier (when I hadn't really looked at the code), I was worried about clearing the `LD_PRELOAD` env. var. for the gzip process; it turns out we are doing it already in the few lines before the exec call.

**Comment by rohgarg** [comment]
Closed via #375.
